### ISSUE TITLE
axis interface cleanup

### ISSUE THREAD

**Issue by HDembinski** [scikit-hep/boost-histogram#96]
- [x] remove axis.bins(...), make axis iterable instead
- [x] expose bin objects, or yield tuples/values? I think yielding tuples/values is better.
- [x] nice `__repr__` for each axis 
- [x] remove axis.update(...) for growing axis [it is not to be called by users and there is no need to unit test this in Python as it is already unit tested in C++]

**Comment by HDembinski** [comment]
Ok, then we agree. We don't expose the C++ bin objects, but instead we provide .edges, .centers, .widths.

**Comment by henryiii** [comment]
And a nice repr, I think that will work well.

**Comment by HDembinski** [comment]
Ok, I see why you didn't have the flow argument for `.centers()`. A center would be meaningless for a bin from [X, infinity)

**Comment by henryiii** [comment]
(Yes, that is why)

**Comment by henryiii** [comment]
> remove axis.update(...) for growing axis

This was requested on Gitter by someone who needed to use it.

> nice `__repr__` for each axis

This is two-part, I've just about finished the axis `__repr__`, but we also need some nice way to show the bins. (Maybe the tuples will be enough here, let's see)

**Comment by HDembinski** [comment]
> This was requested on Gitter by someone who needed to use it.

What for? I am strongly against this and unless it is clear to me that this is a legitimate use, I will assume that this person is just confused about how the library works.

**Comment by henryiii** [comment]
Please go to the Gitter channel and read the messages, starting with Mar 19 17:35. You were active in that discussion.

**Comment by HDembinski** [comment]
If you remember the argument, why don't you post it here? I don't need to convince you, I need to be convinced. I know how users should use boost::histogram and by extension boost-histogram.

Having `axis.update` exposed is only causing confusion. Calling update on an axis that is part of a histogram is not allowed by boost::histogram. The update method requires a mutable axis, while the axis returned by a histogram is const. If `axis.update` is exposed in Python, you need to raise an error when a user calls it on an axis of a histogram. Exposing a method that just raises an error is stupid. While it is technically possible to use a non-const axis independently of the histogram, there are no good use cases for that.

**Comment by henryiii** [comment]
@TaiSakuma was the one that was asking about it. I don't need it and am happy to remove it unless it is truly useful and needed (and yes, it should not be allowed on an axis attached to a histogram).

**Comment by henryiii** [comment]
@TaiSakuma feel free to explain why it is or would be useful if you need it!

**Comment by TaiSakuma** [comment]
@henryiii , I'll write when I can.

**Comment by HDembinski** [comment]
I was thinking more about this. I am trying to find an interface that does not divert too much from C++, gives the right nudge to use the axis performantly in Python, and which allows one to treat discrete and continuous axes the same as much as possible in typical use scenarios.

Proposal:
- remove `.bin(index)`, only keep `.bins(flow=false)`
- remove `.values(flow=false)` for discrete axes, its role is taken over by `.bins(flow=false)`
- make `.centers` a property that returns a numpy array and have it for continuous *and* discrete axes (and cache the numpy array, so it doesn't have to generated again and again)
- remove `.size`, only keep `.__len__(self)` and `.extent`
- offer `.edges(flow=false)` for all axes except for the category axis; it make sense to offer it for the integer axis although it is discrete

**Comment by henryiii** [comment]
1. `.bins()[3]` is not too bad, is less to remember, and most of the time, users will probably want the whole array of bins. It's a little wasteful if a user really does call this multiple times, but again, `b = bins(); ... b[3]` is fine.
2. Consistency and simplicity is good
3. Good, though not sure you need to cache here. It should be fairly fast to generate; you'll need to copy it each call anyway because it should not be modifiable, which I expect will take nearly as long. And adding the caching adds complexity.
4. `size` is pretty common (numpy arrays have it), and `len(a) == len(tuple(a))` must be true or you'll have very confused users. Does iteration include the flow bins? If so, then maybe removing extent would be better (and supporting both len and size is how numpy works). How do you turn on/off the flow bins for iteration, actually?
5. I like `edges` *slightly* better than `np_bins`, and numpy calls the edges array "edges" in the docs. However, the discreet axis case is a little nicer as `np_bins`, but I think it's better to optimize the name for the case it's intended for (the other is just a fallback). But it's a slight preference only.

**Comment by HDembinski** [comment]
> Good, though not sure you need to cache here. It should be fairly fast to generate; you'll need to copy it each call anyway because it should not be modifiable, which I expect will take nearly as long. And adding the caching adds complexity.

Allocating a new numpy array each time is costly. Filling the array is very fast, sure. I don't know what you mean with "you'll need to copy it each call anyway". You can set the numpy array to read-only to make it non-modifiable. For the cache, I think all you need is to check whether the axis size is still the same as the centers array size, which is a very fast check.

> `len(a) == len(tuple(a))` must be true

Yes, that was the idea. I am also ok with using `.size` instead of `.extent`. Then `.size` would represent the physical size and `__len__` the logical size (when ignoring flow bins).

Regarding 5, I like the name `edges` in principle, but there is no way to use `edges` also for a category axis, while it is ok to use `np_bins` for all axis, although it returns something different for a category axis.

**Comment by HDembinski** [comment]
> `.bins()[3]` is not too bad, is less to remember, and most of the time, users will probably want the whole array of bins. It's a little wasteful if a user really does call this multiple times, but again, `b = bins(); ... b[3]` is fine.

Yes, the idea here is to nudge users to access the bins as an array instead of calling `.bin(idx)`.

**Comment by HDembinski** [comment]
I try to make the interface easy to use correctly and difficult to use incorrectly.

**Comment by henryiii** [comment]
> Allocating a new numpy array each time is costly

I would start with the simpler solution, then add this later as an optimization (with a check to make sure it is worth it). Yes, an immutable numpy array would avoid the copy, I didn't think about that till after I responded. That would probably mean the cached version would be quite a bit faster. A user can always manually cache, though, with `b = a.bins()` if performance is important, so I'm not sure if optimizing here is worth the complexity increase. Growing axis would be the one case where it would be better done at this level, but again, not sure how important the optimization would be.

> Yes, that was the idea. I am also ok...

I checked, it looks like iteration over the axis `tuple(a)` does not include the flow bins. So then `__len__` should not include the flow bins. Then it's whether you want to be Numpy-like and use `size` or follow Boost.Histogram and use `extent`. I'm fine with either, and lightly lean toward the latter. Size is not a tuple, so it's not 1:1 to numpy, and enforcing the meaning of extent is good.

What about `a[idx]`? That is something that is commonly expected to work if something is iterable in Python (there are exceptions though, like generators).

> `edges`

I think having a fallback of `edges` returning the categories when there are no edges makes sense, and I don't think I'd intuitively know what `np_bins` were, but it's only a very mild preference, I wouldn't complain about `np_bins`. Maybe ask a few people what they would expect each to return in both cases?

**Comment by HDembinski** [comment]
About `a[idx]`, I like it in principle and would like to have it so the axis has the full PySequence interface, but what do we do with `a[-1]`? In Python this normally means "access the last bin", but in boost::histogram it means "access the underflow". We can of course document this departure from the norm, but we could also try to avoid the issue altogether by not offering `__getitem__`. I haven't made up my mind about this and am interested in opinions.

**Comment by HDembinski** [comment]
Ok, no caching of centers for now. You are right, it can be added later.

**Comment by henryiii** [comment]
I would match the Python/Numpy behavior (since you've now defined iteration to not include flow bins). So `a[-1] == np.asarray(a)[-1]`. Users can still use `.bins(flow=True)`. Maybe we can eventually add a special type with an api (`loc(x)` should work here too, not just in histograms, so we could have `flow(i)` or something like that).

**Comment by HDembinski** [comment]
I am inclined to agree, although I am not happy to have Python behave different from C++. The indexed iterator also needs to return shifted indices then when the user requests iteration over all bins.

**Comment by HDembinski** [comment]
Just as a reminder, our guiding idea was to hide the flow bins unless the user explicitly requests to see them.

**Comment by HDembinski** [comment]
It therefore follows that normal iteration over an axis or histogram should skip them.

**Comment by HDembinski** [comment]
There is a way to have the cake and eat it, namely the trick that I used before in the old Python bindings. If `.bins(True)` returns [bin0, ..., binN-1, binOverflow, binUnderflow], then the index -1 can mean the same thing in C++ and Python. The caveat then is that `.edges(True)` cannot be implemented. But I think it is important that Python and C++ are consistent.

**Comment by HDembinski** [comment]
I would rather sacrifice `.edges(True)` than the consistency between Python and C++. One can still have `.edges(False)`

**Comment by HDembinski** [comment]
We have to keep the C++ convention, otherwise the consistency breaks everywhere. The convention is that axis.index returns -1 when the value is in the underflow. For an axis with underflow bin, this has to address a valid bin. `axis[-1]` should return the underflow bin or raise an IndexError when it doesn't exist.

**Comment by HDembinski** [comment]
We have to document that the index -1 is used differently. This was my original solution and I thought long and hard about it. Any solution needs to take the consistency of the whole library into account. The meaning of the indices returned by `axis.index` shouldn't change whether or not the axis has underflow bins. This was one of the core design principles in boost::histogram and boost-histogram may not divert from that.

**Comment by henryiii** [comment]
I think that having consistency with the UHI (Unified Histogram Indexing) proposal is important as well. You could have tag types; `bh.underflow` and `bh.overflow`, which would simply implement an attribute `x.overflow=True` or `x.underflow=False`. If the value passed into `[]` had a `*flow` attribute, then it returns that bin. This could *also* be used in the histogram indexing, such that `h[bh.underflow:bh.overflow:bh.project]` would be the same thing as `h[::bh.project]`, and `h[bh.underflow]` would access the underflow bin, etc.

This way we don't have to break consistency with the Python conventions. We are not implementing `.bin(i)` from what I gathered, so we are also not breaking consistency with Boost.Histogram - if we implemented `.bin(i)`, then `.bin(-1)` should be the underflow bin.

**Comment by henryiii** [comment]
I can extend the UHI proposal to include underflow/overflow if you want to see it.

Also, my UHI code in boost-histogram may get rewritten; I'd like to see if it looks cleaner in Python.

I'd recommend leaving off `[]` access and just working on the rest until we get a bit further on UHI for histograms.

**Comment by HDembinski** [comment]
Can you post a link for Unified Histogram Indexing here?

Even if we don't implement `.bin(i)`, we have `.bins()[i]` and the `i` could come from `axis.index`.

**Comment by HDembinski** [comment]
You seem to not value the boost::histogram convention very much, but I do for obvious reasons.

**Comment by HDembinski** [comment]
I suppose both the UHI and the boost::histogram convention can be implemented in parallel. If I remember correctly, `bh.underflow` is not an integer.

**Comment by henryiii** [comment]
`bh.underflow` was not part of the UHI proposal yet, but it should be added (and should not be an integer, but a type with an attribute.

**Comment by henryiii** [comment]
A user may try `[~0]` to get the final bin of a histogram (or, more likely, `[-1]`, but I like the symmetry of starting from 0 better, and in Python inversion is always correctly defined), and will be very confused if that is not the final bin of the histogram they thought they had. And if they loop over bins, they will be very surprised if they end up with overflow/underflow bins, especially if the order is backward (which will never happen in C++, since it's not tacking the underflow bin on the end). And if `[]` does not match looping, that would be very odd indeed!

**Comment by HDembinski** [comment]
So your proposal is to have `bins()` return [bin0, binN-1, overflow, underflow] which is consistent boost::histogram indexing, but `axis[idx]` would follow Python conventions.

**Comment by henryiii** [comment]
So I would very strongly recommend that `[~i]` and `[-i]` behave exactly like Python users expect, and we can add the `bh.under/overflow` tags to access the bins.

**Comment by henryiii** [comment]
I was thinking `bins()` would return `[bin0,...,binN]` and `bins(flow=True)` would return `[underflow,...,overflow]`, and we should not try to reorder them to match indexing. If you would like boost.histogram style indexing, you could add `.bin(i)` which would exactly match Boost.Histogram. The actual Boost.Histogram does not have `.bins()`.

**Comment by HDembinski** [comment]
But how do you handle the inconsistency then, that you cannot use the indices returned by `axis.index` properly in `axis[]`. That is confusing.

**Comment by henryiii** [comment]
Most people will either be looping over the returned value from `bins()`, or they will be feeding it to numpy, neither of which will expect oddly ordered flow bins.

**Comment by henryiii** [comment]
Could `axis.index` return underflow/overlfow instead of `-1` and `N-1+1`? Haven't thought about that one much.

**Comment by HDembinski** [comment]
Yes, if you give up on returning a numpy array of indices.

**Comment by henryiii** [comment]
I told you I hadn't thought about it much. It could be directly used in `.bin()`, if it returned `-1` and `N-1+1`.

**Comment by HDembinski** [comment]
Then please think about it some and tell me whether you see a consistent solution other than mine.

**Comment by henryiii** [comment]
I still like the idea of also supporting the tag types, but disallowing negative access to contents and using -1 doesn't sound as bad then. Let me think a bit.

**Comment by HDembinski** [comment]
I also like to support tag types in addition, I think that can work.

**Comment by henryiii** [comment]
I do still think `bins()` should return standard order Python objects.

**Comment by HDembinski** [comment]
It would return standard order for `.bins(flow=False)`

**Comment by henryiii** [comment]
By the way, here's the link: https://boost-histogram.readthedocs.io/en/latest/usage/indexing.html

**Comment by henryiii** [comment]
It returns a standard numpy object, and those should respect standard conventions. Since axis is our object, we can support custom access.

**Comment by HDembinski** [comment]
If you want this to be *universal* histogram indexing, it may be good to have an extra web page where this idea is presented.

**Comment by henryiii** [comment]
It will show url in a different place eventually, but currently I don't have another webpage. I think we could move this to Scikit-HEP's website when we are ready. (It still may change a bit, and boost-histogram is the only implementation so far).

**Comment by henryiii** [comment]
(I'm also currently marking what works and doesn't work, which would not be in universal docs :) )

**Comment by HDembinski** [comment]
The following code should work correctly (1d histogram): `axis[axis.index(-inf)]` and `axis.bins()[axis.index(-inf)]`.

**Comment by HDembinski** [comment]
For a single argument to `axis.index` we could make it return a tag, but not if people use the vectorized form of `axis.index`.

**Comment by HDembinski** [comment]
To keep the usual Python meaning of `-1`, the underflow bin has to come "last". It is a little unusual, but the best solution I can think of.

**Comment by henryiii** [comment]
Why have two ways to do the same thing? And that makes *the default* behavior to add the overflow/underflow bins, which this starts failing: `len(axis.bins()) == len(axis)` along with the more important one, iteration over bins vs. iteration over axis. I think the return value for bins() should behave nicely everywhere else, and should default (like everything, for consistency) to not having the flow bins added. Maybe we can have a `flow=bh.cppstyle` option. :) (or, more realistically, flow could be set with a enum)

**Comment by henryiii** [comment]
I think most Python users will not want to interact directly with flow - it mostly helps things like projections to work reasonably. Many of the ones that do actually want the flow bins will want them to feed into something else; nothing else in the Python ecosystem expects the out of order pattern like this. The few that want to match Boost.Histogram's ordering could use a special version of `flow=`, perhaps?

**Comment by henryiii** [comment]
You can have underflow/overflow with numpy, for example, but it will not reorder like this.

**Comment by HDembinski** [comment]
I care strongly about indices behaving consistently in this library.

**Comment by henryiii** [comment]
```
>>> np.histogram([-1,2,3,4,5], [-np.inf, -1, 0, 1, np.inf] )
(array([0, 1, 0, 4]), array([-inf,  -1.,   0.,   1.,  inf]))
```

You want boost-histogram to return `(array([1, 0, 4, 0]), array([-1.,   0.,   1.,  inf, -inf]))`. You yourself have said that most users will be familiar with Python and not Boost.Histogram. Breaking this is something we should not do by default, IMO. Maybe as a special option.

**Comment by HDembinski** [comment]
That is not what I want. You can't return an edge array, like I said above.

**Comment by HDembinski** [comment]
An edge array can only be returned with `flow=False`
